placement/algo/helper.go

```diff
@@ -78,192 +82,216 @@ func (ph *placementHelper) MoveShard(shard uint32, from, to placement.HostShards
+		for _, triedHost := range triedHosts {
+			heap.Push(hostHeap, triedHost)
+		}
+		triedHosts = triedHosts[:0]
 	}
 	return nil
 }
 
-func (ph *placementHelper) GetHostHeap() heap.Interface {
-	return ph.hostHeap
-}
+// placeToRacksOtherThanOrigin move shards from a host to the rest of the cluster
+// the goal of this function is to assign "some" of the shards to the hosts in other racks
+func (ph placementHelper) placeToRacksOtherThanOrigin(shards map[uint32]int, from placement.HostShards) {
+	var otherRack []placement.HostShards
+	for rack, hostShards := range ph.rackToHostsMap {
+		if rack == from.Host().Rack() {
+			continue
+		}
+		for hostShard := range hostShards {
+			otherRack = append(otherRack, hostShard)
+		}
+	}
 
-func (ph *placementHelper) GenerateSnapshot() placement.Snapshot {
-	return placement.NewPlacementSnapshot(ph.hostShards, ph.uniqueShards, ph.rf)
+	hostHeap := ph.BuildHostHeap(otherRack, true)
+
+	var triedHosts []placement.HostShards
+	for shard := range shards {
+		for hostHeap.Len() > 0 {
+			tryHost := heap.Pop(hostHeap).(placement.HostShards)
+			if ph.GetTargetLoadForHost(tryHost.Host().ID())-tryHost.ShardsLen() <= 0 {
+				// this is where "some" is, at this point the best host option in the cluster
+				// from a different rack has reached its target load, time to break out of the loop
+				return
+			}
+			triedHosts = append(triedHosts, tryHost)
+			if ph.MoveShard(shard, from, tryHost) {
+				delete(shards, shard)
+				break
+			}
+		}
+
+		for _, triedHost := range triedHosts {
+			heap.Push(hostHeap, triedHost)
+		}
+		triedHosts = triedHosts[:0]
+	}
 }
 
 // NewPlacementHelper returns a placement helper
-func NewPlacementHelper(opt placement.Options, s placement.Snapshot) PlacementHelper {
-	return newPlacementHelperWithTargetRF(opt, s, s.Replicas())
+func NewPlacementHelper(ps placement.Snapshot, opt placement.Options) PlacementHelper {
+	return newHelper(ps, ps.Replicas(), opt)
 }
-func newInitPlacementHelper(opt placement.Options, hosts []placement.Host, ids []uint32) PlacementHelper {
+
+func newInitHelper(hosts []placement.Host, ids []uint32, opt placement.Options) PlacementHelper {
 	emptyPlacement := placement.NewEmptyPlacementSnapshot(hosts, ids)
-	return newPlaceShardingHelper(opt, emptyPlacement, emptyPlacement.Replicas()+1, true)
+	return newHelper(emptyPlacement, emptyPlacement.Replicas()+1, opt)
 }
 
-func newPlacementHelperWithTargetRF(opt placement.Options, s placement.Snapshot, targetRF int) PlacementHelper {
-	return newPlaceShardingHelper(opt, s, targetRF, true)
+func newAddReplicaHelper(ps placement.Snapshot, opt placement.Options) PlacementHelper {
+	return newHelper(ps, ps.Replicas()+1, opt)
 }
 
-func newAddHostShardsPlacementHelper(opt placement.Options, s placement.Snapshot, hs placement.HostShards) PlacementHelper {
-	var hss []placement.HostShards
+func newAddHostShardsPlacementHelper(ps placement.Snapshot, hostShards placement.HostShards, opt placement.Options) PlacementHelper {
```

Comment:
Maybe we should rename this too from `newAddHostShardsPlacementHelper` to just `newAddHostShardsHelper`?